### PR TITLE
update: add helm test step

### PR DESCRIPTION
**Pull request by charles-woshicai**
Create an action to test helm charts, as required in https://github.com/oam-dev/kubevela/issues/1406

### CODE REVIEW

**Comment by charles-woshicai** [comment]
Add an action by following the doc of [this page](https://github.com/marketplace/actions/helm-chart-testing-helm-3)

**Comment by charles-woshicai** [comment]
Use chart-test-action to test charts, [here](https://github.com/helm/chart-testing-action) is the docs.

**Comment by charles-woshicai** [comment]
Review needed @wonderflow .

File: .github/workflows/lint-test.yaml
Comment (wonderflow) [inline .github/workflows/lint-test.yaml:36]: what is the comand `ct` ?

File: .github/workflows/lint-test.yaml
Comment (wonderflow) [inline .github/workflows/lint-test.yaml:54]: It should have a `helm test` runing here to validate, I didn't see them?

https://helm.sh/docs/topics/chart_tests/

File: charts/vela-core/templates/test/test-application.yaml
Comment (wonderflow) [inline charts/vela-core/templates/test/test-application.yaml:19]: As the docs said, the test should be a K8s job with command running.  https://helm.sh/docs/topics/chart_tests/

**Review by wonderflow — CHANGES_REQUESTED** [review]
We could add the chart test in `e2e-test` instead of  creating new one

File: .github/workflows/lint-test.yaml
Comment (charles-woshicai) [reply .github/workflows/lint-test.yaml:36]: `ct` is equal to `chart testing`, a command-line tool which is maintained by `helm`.

[Here](https://github.com/helm/chart-testing) is the repository of `ct`.

File: .github/workflows/lint-test.yaml
Comment (charles-woshicai) [reply .github/workflows/lint-test.yaml:54]: `ct install` would run both helm `install` and `test`.

[Here](https://github.com/helm/chart-testing/blob/master/doc/ct_install.md#ct-install) is the doc ref.

File: charts/vela-core/templates/test/test-application.yaml
Comment (charles-woshicai) [reply charts/vela-core/templates/test/test-application.yaml:19]: > We could add the chart test in `e2e-test` instead of creating new one

Now we detect whtehre there is a change in charts via `ct list-change`, which comares the charts between `master` branch and `PR` branch. If the charts is not changed,  there is no necessary to run `helm install & helm test`. 

Is it necessary to run `helm install & helm test` in `e2e-test`?

> As the docs said, the test should be a K8s job with command running. https://helm.sh/docs/topics/chart_tests/

I will fix it when I am free.

Comment (wonderflow) [reply charts/vela-core/templates/test/test-application.yaml:19]: > Is it necessary to run helm install & helm test in e2e-test?

I think we can always run the check. Does it cost too much time in CI?

File: .github/workflows/lint-test.yaml
Comment (wonderflow) [reply .github/workflows/lint-test.yaml:54]: I see, thanks, but use a yaml about `Application` seems not check the check well.

Comment (wonderflow) [reply .github/workflows/lint-test.yaml:54]: We should at least check the Application is running and has a deployment and ingress created

Comment (charles-woshicai) [reply .github/workflows/lint-test.yaml:54]: I agree, an application just test whether crd is installed.

Will fix it.

File: charts/vela-core/templates/test/test-application.yaml
Comment (charles-woshicai) [reply charts/vela-core/templates/test/test-application.yaml:19]: > > Is it necessary to run helm install & helm test in e2e-test?
> 
> I think we can always run the check. Does it cost too much time in CI?

I review the charts and it makes sense to run check every time for the image tag of core controller is `latest` in the master branch.

Will adopt to move charts check to `e2e-test`.

File: .github/workflows/go.yml
Comment (wonderflow) [inline .github/workflows/go.yml]: the kind already setup in L174

File: .github/workflows/go.yml
Comment (wonderflow) [inline .github/workflows/go.yml]: why need to apply this file?

Comment (charles-woshicai) [reply .github/workflows/go.yml]: FYI the code is not the final state now.

I want to check the `PHASE` and `HEALTH` of `applicatoin` to be `running` and `healthy` in test pod.
In my local environment the `HEALTH` of application stuck in `checkHealth`, this line is just having a try. If it's no necessary, I will  remove it.

**Comment by charles-woshicai** [comment]
Review needed. @wonderflow 
I just chcek whether vela-controller created `depoloy` and `ing` without waiting for `application` ready.

Could I rebase total commits in this PR to one?  There  are too many debug commits, which is useless.

**Comment by wonderflow** [comment]
> BTW, could I rebase total commits in this PR to one? There are too many debug commits, which is useless.

Sure, please go ahead. I can also help squash before merge.

File: .github/workflows/go.yml
Comment (wonderflow) [inline .github/workflows/go.yml:173]: please check the `make e2e-setup` command, in the makefile, it has already install the helm chart. We can just merge `helm lint` and `helm test` into the `e2e-setup`.

Comment (wonderflow) [inline .github/workflows/go.yml:173]: BTW, why check the logs?

File: charts/vela-core/templates/test/test-application.yaml
Comment (wonderflow) [inline charts/vela-core/templates/test/test-application.yaml:57]: will the `helm test` fail if the job is hanging on sleep? Do we have any timeout settings?

File: .github/workflows/go.yml
Comment (wonderflow) [inline .github/workflows/go.yml:153]: please clean these unused code

File: .github/workflows/go.yml
Comment (charles-woshicai) [reply .github/workflows/go.yml:173]: It's not necessary, I will remove it.

Comment (charles-woshicai) [reply .github/workflows/go.yml:173]: OK, I will take a look.

File: charts/vela-core/templates/test/test-application.yaml
Comment (charles-woshicai) [reply charts/vela-core/templates/test/test-application.yaml:57]: Yeah the timeout is `5m` for now.

Comment (wonderflow) [reply charts/vela-core/templates/test/test-application.yaml:57]: That's good

**Comment by charles-woshicai** [comment]
I have rebased all commits into one.

**Review by wonderflow — APPROVED** [review]
LGTM after CI passed

**Comment by wonderflow** [comment]
@charles-woshicai great job, thanks for your contributions!

**Pull request merged**
